shards/sched.go

```diff
@@ -11,7 +15,44 @@ import (
 // Note: This is a Sourcegraph specific addition to allow long running queries
 // along normal interactive queries.
 
-// scheduler is for managing concurrent searches. Its goals are:
+// scheduler is for managing concurrent searches.
+type scheduler interface {
+	// Acquire blocks until a normal process is created (ie for a search
+	// request). See process documentation. It will only return an error if the
+	// context expires.
+	Acquire(ctx context.Context) (*process, error)
+
+	// Exclusive blocks until an exclusive process is created. An exclusive
+	// process is the only running process. See process documentation.
+	Exclusive() *process
+}
+
+// The ZOEKTSCHED environment variable controls variables within the
+// scheduler. It is a comma-separated list of name=val pairs setting these
+// named variables:
+//
+//   disable: setting disable=1 will use the old zoekt scheduler.
+//
+// Note: these tuneables should be regarded as temporary while we experiment
+// with our scheduler in production. They should not be relied upon in
+// customers/sourcegraph.com in a permanent manor (only temporary).
+var zoektSched = parseTuneables(os.Getenv("ZOEKTSCHED"))
```

Comment:
This is a really nice pattern!

Reply:
Thanks. Yeah I was adding a bunch of envvars, then thought this has got to be better. Then thought of how the go runtime does it :D